### PR TITLE
dir.c: regression fix for add_excludes with fscache

### PR DESCRIPTION
**Pull request by jeffhostetler**
Fix regression described in:
https://github.com/git-for-windows/git/issues/1392
which was introduced in:
https://github.com/git-for-windows/git/commit/b2353379bba414e6c00dde913497cc9c827366f2

When the user has a .gitignore file that is a symlink, the fscache
optimization introduced above caused the stat-data from the symlink,
rather that of the target file, to be returned.  Later when the ignore
file was read, the buffer length did not match the stat.st_size field
and we called die("cannot use <path> as an exclude file")

Added code to replace the stat-data when the path is a symlink.  This
preserves the effect of the performance gains for the normal case where
it is not a symlink.

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

### CODE REVIEW

File: dir.c [hunk 1]
```diff
@@ -758,13 +758,33 @@ static int add_excludes(const char *fname, const char *base, int baselen,
...
+		 *
+		 * We cannot use the fscache-stat helper because it needs to
+		 * actually open the file to dereference symlinks and then do
+		 * fstat.  This defeats our purpose here.
+		 *
+		 * If the .gitignore file is successfully opened and in the
+		 * rare case that it is a symlink, we fstat it to replace the
+		 * original lstat stat-data to complete our illusion.
+		 */
 		if (lstat(fname, &st) < 0) {
```
Comment (dscho) [inline dir.c]: I wonder if all that is needed is an `s/lstat/stat/`?

File: compat/win32/fscache.c [hunk 2]
```diff
@@ -257,7 +252,7 @@ static void fscache_clear(void)
 /*
  * Checks if the cache is enabled for the given path.
  */
-static inline int fscache_enabled(const char *path)
+int fscache_enabled(const char *path)
```
Comment (dscho) [inline compat/win32/fscache.c]: This hunk and the one above look like a good change in its own right. Maybe that merits a separate commit, too?

File: dir.c (hunk 1 above)
Comment (jeffhostetler) [reply dir.c]: no. see updated commit message.

**Comment by dscho** [comment]
Thank you so much!

**Pull request merged**